### DISCUSSION TITLE
What is keg-only?

### DISCUSSION THREAD

**Discussion by mudasar187** [Homebrew/discussions#239] [Everyday usage] [answered]
Hi!

Sometime when I update my applications trough brew i get following:

```
sqlite is keg-only, which means it was not symlinked into /usr/local,
because macOS already provides this software and installing another version in
parallel can cause all kinds of trouble.

If you need to have sqlite first in your PATH run:
  echo 'export PATH="/usr/local/opt/sqlite/bin:$PATH"' >> ~/.zshrc

For compilers to find sqlite you may need to set:
  export LDFLAGS="-L/usr/local/opt/sqlite/lib"
  export CPPFLAGS="-I/usr/local/opt/sqlite/include"
```

What does this means? Should I add these?

```
If you need to have sqlite first in your PATH run:
  echo 'export PATH="/usr/local/opt/sqlite/bin:$PATH"' >> ~/.zshrc

For compilers to find sqlite you may need to set:
  export LDFLAGS="-L/usr/local/opt/sqlite/lib"
  export CPPFLAGS="-I/usr/local/opt/sqlite/include"
````

And should I do it for all applications that tell me to do it? 

Thank you!

**Comment by carlocab** [comment, marked as answer]
To understand what's going on, you need to understand the following terminology: keg, `PATH`, and symlink. The following links may help:

Q: What is a keg?
https://docs.brew.sh/Formula-Cookbook#homebrew-terminology

Q: What is `PATH`?
https://www.cs.purdue.edu/homes/bb/cs348/www-S08/unix_path.html
https://en.wikipedia.org/wiki/PATH_(variable)

Q: What is a symlink?
https://en.wikipedia.org/wiki/Symbolic_link

The way Homebrew works is that it installs packages into your cellar (see link above on Homebrew terminology), and then symlinks it into `/usr/local` (in a default Homebrew installation). It sometimes forgoes the symlinking step when it has good reasons to, and it tells you when this happens. The packages for which the symlinking step is skipped are called keg-only. If you want to override Homebrew's decision not to symlink a package into `/usr/local`, you can modify your `PATH` variable as in the suggestion from `brew` that you've posted.

Only you can decide whether you need to override the fact that a package is keg-only. If you're not sure, it's probably best not to.

As for `LDFLAGS` and `CPPFLAGS`, it's best to ignore them if you don't know what they are. If you do, then you'll know what you need to do with them. These are essentially ways to give instructions to a compiler when you are compiling software. If you do not build software from source, then you don't need to worry about this.

**Reply by gromgit** [reply]
@carlocab has summarized the situation quite well, so I'll just add one thing:

> The way Homebrew works is that it installs packages into your cellar (see link above on Homebrew terminology), and them symlinks it into `/usr/local` (in a default Homebrew installation).

There's at least one more step in between. From the user's POV, when installing a formula (say, `sqlite`), Homebrew:
1. installs the `sqlite` formula's files under `/usr/local/Cellar/sqlite/<sqlite_ver>`
1. symlinks `/usr/local/Cellar/sqlite/<sqlite_ver>` as `/usr/local/opt/sqlite`
1. symlinks the contents of `/usr/local/Cellar/sqlite/<sqlite_ver>` into `/usr/local`

For keg-only formulae, step 3 is omitted. Step 2 is done so that you can always access the "active" version of `sqlite` via `/usr/local/opt/sqlite`, even for keg-only formulae. For instance, the Homebrew-installed `sqlite3` CLI binary is always available via `/usr/local/opt/sqlite/bin/sqlite3`, regardless of version, and can therefore be safely hardcoded into your scripts.

As a general rule, keg-only formulae should _not_ be forced into the main Homebrew hierarchy via `brew link`, for the reasons documented in the formulae. There's almost always a better way to use such formulae...also documented in the formulae.

**Comment by dexterous** [comment]
Is there a way to explicitly install a formula as keg-only? Or is this something only a Formula author can control?

**Reply by SMillerDev** [reply]
If you `brew unlink <formula>` it will become keg-only. This will not survive updates though.

**Reply by dexterous** [reply]
Yeah, but I would've preferred a workflow where it didn't link to start with. I did try it with `--no-binaries` as it seemed like the closest best alternative. However, luckily, since I was playing with `docker`, it refused to link anyway because I had the Cask already installed and the Formula conflicted with that.

**Reply by SMillerDev** [reply]
Keg-only does not exist for casks. Or do you mean that you wouldn't want any formula to ever link?

**Reply by dexterous** [reply]
Apologies for the confusion, I didn't mean to imply that I was looking for a Keg-only option for Casks. I meant that I wanted to see if there was an option to install any given Formula as Keg-only. Something like...
```shell-session
$ brew install --formula --keg-only docker
```
... which, I think, is equivalent to...
```shell-session
$ brew install --formula docker && brew unlink docker
```
... but I just wanted to avoid using 2 commands.

I hope that clarifies my original question.
